Make check_type independent of global config

antonagestam opened the pull request:
Addresses #316.

antonagestam commented on src/typeguard/_functions.py:
```diff
@@ -71,18 +83,21 @@ def check_type(value: object, expected_type: Any) -> Any:
-def check_argument_types(memo: CallMemo) -> Literal[True]:
+def check_argument_types(
+    memo: CallMemo,
+    config: TypeCheckConfiguration = global_config,
```
> My own changes, which I scrapped earlier, contained the configuration in the memo.

Let me know which way you want to go, I have no strong opinion either way.

> What's the point of adding this parameter to these functions, when it's never being passed any explicit value?

I thought it made sense for these to be in-line with the signature of `check_type`. The version history says these functions are removed, should they be?

agronholm replied:
> I thought it made sense for these to be in-line with the signature of check_type. The version history says these functions are removed, should they be?

They're only used internally now, by both `@typechecked` and the import hook.

antonagestam replied:
I've restored these changes, addressing `@typechecked` and the import hook to not use global state could be done separately from this PR.

antonagestam commented:
@agronholm I addressed your comment about unnecessarily changing the internal functions, I agree that makes sense. I do think it will make sense to address the decorator and import hook in a future PR though, my opinion is that relying on a globally shared and mutable configuration is a recipe for unpleasant surprises, but again, that's opinion.

I made another change here in that I unpacked the configuration parameters into explicit arguments for `check_type`, instead of taking a single configuration object. This is also opinionated, my reasoning is that it's a bit simpler API, less work and import on the call sites of `check_type`. Let me know if you're not on board with this, and I'll happily roll back to taking a `TypeCheckConfiguration` there instead.

A third change, which makes this PR a breaking change, is that the new arguments of `check_type` now default to the _default values of `TypeCheckConfiguration`_, and __not to the values in `global_config`__. Again, this is my opinion about globally shared state and surprises, I just think it makes sense for the defaults of this function to be static.

Docs and tests are updated, and this PR is ready for a re-review.

agronholm commented on src/typeguard/_functions.py:
```diff
@@ -61,21 +91,32 @@ def check_type(value: object, expected_type: Any) -> Any:
     :raises TypeCheckError: if there is a type mismatch
 
     """
+    config = TypeCheckConfiguration(
+        forward_ref_policy=forward_ref_policy,
+        typecheck_fail_callback=typecheck_fail_callback,
+        collection_check_strategy=collection_check_strategy,
+    )
```
This overrides all of the configuration, even if the user just wants to override a single option.

antonagestam replied:
Not really, the defaults gathered from the defaults from of `TypeCheckConfiguration`: https://github.com/agronholm/typeguard/blob/2f4c104770adcc3b0d53bbc52a7f05bf787808b2/src/typeguard/_functions.py#L71

But the semantics are different from `@typechecked` and import hook, in that this is completely detached from the global config. My reasoning is that `check_type` is likely to be used differently from those, like in my case with phantom-types, and in this context it doesn't make sense to have shared mutable state. Does that make sense? :)

agronholm replied:
> Not really, the defaults gathered from the defaults from of `TypeCheckConfiguration`:

The issue is that it disregards the current global config entirely, unlike how it works in `master`. This is one breaking change I'm not sure I want to merge.

antonagestam replied:
That's correct, as I mentioned in the comment above, I want to challenge the choice to make this function's behavior depend on shared mutable state. I understand why it makes sense for `@typechecked` and the import hook to have a global config object, but I think it's unwise for this function to depend on it.

Consider if these field are made to default to the global config, in order to fulfill the case where you don't want to be affected by the globally shared state, then you would always have to pass all configuration. What's worse, if typeguard adds more configuration options to this function in the future, it would be very hard for a user to detect that these have been added and that they should update all their calls to include a static default value.

I.e., my opinion is that `check_type` is sufficiently different from the instrumenting APIs to be treated differently.

agronholm replied:
Alright, you're starting to convince me. This change would have to be reflected in the documentation too then. And `@typechecked` needs to have the overrides added to it too.

antonagestam commented on src/typeguard/_functions.py:
```diff
@@ -36,16 +41,41 @@
+    value: object,
+    expected_type: Any,
+    *,
+    forward_ref_policy: ForwardRefPolicy = TypeCheckConfiguration.forward_ref_policy,
```
Should probably create a default instance of `TypeCheckConfiguration` instead of referencing class attributes, because this would most likely break if a definition is using `dataclasses.field()`.

agronholm replied:
Maybe `forward_ref_policy: ForwardRefPolicy = TypeCheckConfiguration().forward_ref_policy` ?

antonagestam replied:
Linters complained about instantiations in defaults, so I did almost as suggested but broke out to a module-local variable: `_default_config`.

agronholm replied:
What was the complaint?

antonagestam replied:
For this diff:

```diff
diff --git a/src/typeguard/_functions.py b/src/typeguard/_functions.py
index db51c1e..66cc92e 100644
--- a/src/typeguard/_functions.py
+++ b/src/typeguard/_functions.py
@@ -73,7 +73,7 @@ def check_type(
     *,
     forward_ref_policy: ForwardRefPolicy = _default_config.forward_ref_policy,
     typecheck_fail_callback: (TypeCheckFailCallback | None) = (
-        _default_config.typecheck_fail_callback
+        TypeCheckConfiguration().typecheck_fail_callback
     ),
     collection_check_strategy: CollectionCheckStrategy = (
         _default_config.collection_check_strategy
```

I get this:

```sh
$ pre-commit run ruff --all-files
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

src/typeguard/_functions.py:76:9: B008 Do not perform function call `TypeCheckConfiguration` in argument defaults
Found 1 error.
```

agronholm replied:
That seems to come from a flake8-bugbear rule. It's safe to ignore. You can change add `"B008"` to `tool.ruff.ignore` in `pyproject.toml` where it already has `"PGH001"`.

agronholm commented on docs/versionhistory.rst:
```diff
@@ -8,6 +8,9 @@ This library adheres to `Semantic Versioning 2.0 <https://semver.org/#semantic-v
 - Fixed ``warn_on_error()`` not showing where the type violation actually occurred
 - Fixed local assignment to ``*args`` or ``**kwargs`` being type checked incorrectly
 - Fixed ``TypeError`` on ``check_type(..., None)``
+- **BACKWARD INCOMPATBILE** Made ``check_type`` independent of ``typeguard.config``, it
+  now has the same default configuration values as ``TypeCheckConfiguration`` and is
+  given the ability to accept configuration options as keyword arguments.
```
Backward incompatible changes should be displayed first.
```suggestion
- **BACKWARD INCOMPATIBLE** Made ``check_type()`` independent of ``typeguard.config``, it
  now has the same default configuration values as ``TypeCheckConfiguration`` and is
  given the ability to accept configuration options as keyword arguments.
- Fixed ``warn_on_error()`` not showing where the type violation actually occurred
- Fixed local assignment to ``*args`` or ``**kwargs`` being type checked incorrectly
- Fixed ``TypeError`` on ``check_type(..., None)``
```

antonagestam replied:
Fixed.

agronholm commented on docs/versionhistory.rst:
```diff
@@ -5,12 +5,12 @@ This library adheres to `Semantic Versioning 2.0 <https://semver.org/#semantic-v
-- Fixed ``warn_on_error()`` not showing where the type violation actually occurred
-- Fixed local assignment to ``*args`` or ``**kwargs`` being type checked incorrectly
-- Fixed ``TypeError`` on ``check_type(..., None)``
 - **BACKWARD INCOMPATBILE** Made ``check_type`` independent of ``typeguard.config``, it
```
Fixed a typo and added () to indicate function.
```suggestion
- **BACKWARD INCOMPATIBLE** Made ``check_type()`` independent of ``typeguard.config``, it
```

antonagestam replied:
This exceeds line-length, I'll push a re-formatting :)

antonagestam replied:
Oh, noticed rest of file doesn't adhere to line-length limit, so I'll leave it :)

agronholm commented:
It looks like you need to do a merge from `master` before this PR becomes mergeable.

antonagestam commented:
@agronholm Is Github being weird somehow? I see this:

<img width="1068" alt="Skärmavbild 2023-03-19 kl  15 11 52" src="https://user-images.githubusercontent.com/218551/226181861-2558854a-7e8c-4fb4-8f7c-3716f122f233.png">

And git reports being up-to-date with master (I synced yesterday, merge commit: [2f4c104](https://github.com/agronholm/typeguard/pull/317/commits/2f4c104770adcc3b0d53bbc52a7f05bf787808b2)):

```sh
$ git fetch -p upstream && git merge upstream/master 
Already up to date.
```

Do you want me to squash, or do you use Github's "Squash and merge" feature?

agronholm commented:
This is what I see:
![image](https://user-images.githubusercontent.com/130003/226188188-cc8e519c-311a-4f09-b7a0-c5d88b14103f.png)

> Do you want me to squash, or do you use Github's "Squash and merge" feature?

I always squash when merging PRs.

agronholm commented:
I now have a new branch for the upcoming v4.0: `4.0-dev`. This PR should target that branch.

agronholm commented:
There's a new checker that needs updating.

agronholm commented:
The failures on py3.7 came from my side. I'll fix them.

antonagestam commented:
I've squashed and rebased on the 4.0 branch, fixed conflicts, and reversed to prefer multiple instantiations of `TypeCheckConfiguration` over a shared instance for defaults, ignoring B008.

agronholm commented:
Sounds great! I'll fix the 3.7 failures either today or tomorrow. Meanwhile, you can add the new parameters to `typechecked()`. The AST transformer code (which also needs to be updated to support this) might look a bit scary, but I can handle that.

agronholm commented:
Do you want me to take over from here?

antonagestam commented:
@agronholm Please go ahead ☺️

agronholm commented:
I'm not yet done with this PR, but this latest changeset should show what I intend to do.

agronholm commented:
Thanks for this. I hope I can release v4.0 soon now.

antonagestam commented:
@agronholm Thanks for bringing this across the finish line 🙏
